### PR TITLE
[query] Support indexed table reads in lowered execution

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
a large diff, but not that complex. The changes to Emit/EmitStream are rote -- adding `ctx` to the EmitStreamContext and entry points.

The design here (serializing the unstaged closures to generate decoders and such) is pretty gross, and I intend to fix this when we have infrastructure to mix interpreted TableIRs and lowered TableStages, and we can remove the interpreted indexed reader.

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 1]
```diff
@@ -488,6 +493,100 @@ case class PartitionNativeReader(spec: AbstractTypedCodecSpec) extends Partition
...
+            k(COption(!hasNext, next))))
+        .map(
+          pc => EmitCode.present(eltType, pc),
+          setup0 = None,
+          setup = Some(
+            EmitCodeBuilder.scopedVoid(mb) { cb =>
+              val ctxMemo = ctxStruct.asBaseStruct.memoize(cb, "pnri_ctx_struct")
+              cb.assign(idxr, getIndexReader(ctxMemo
+                .loadField(cb, "indexPath")
+                .handle(cb, cb._fatal(""))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:547]: You can use `IEmitCode.get` when you don't have a specific error message.

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 2]
```diff
@@ -553,26 +652,33 @@ class TableNativeReader(
...
-      body)
+    val specPart = rowsSpec.partitioner
+    val partitioner = if (filterIntervals)
+      params.options.map(opts => RVDPartitioner.union(specPart.kType, opts.intervals, specPart.kType.size - 1))
+    else
+      params.options.map(opts => new RVDPartitioner(specPart.kType, opts.intervals))
+
+    spec.rowsSpec.readTableStage(ctx, spec.rowsComponent.absolutePath(params.path), requestedType.rowType, partitioner, filterIntervals).apply(globals)
+//
+//    val rowsPath = spec.rowsComponent.absolutePath(params.path)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:664]: Delete commented out old code?

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
Sorry for the delay. Just a couple nits.

**Comment by tpoterba** [comment]
addressed, thanks for the comments.

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**